pybamm/util.py

```diff
@@ -345,3 +345,21 @@ def install_jax(arguments=None):  # pragma: no cover
             f"jaxlib>={JAXLIB_VERSION}",
         ]
     )
+
+
+def have_optional_dependency(module_name, attribute=None):
+    try:
+        module = importlib.import_module(module_name)
+        if attribute:
+            if hasattr(module, attribute):
+                imported_attribute = getattr(module, attribute)
+                return imported_attribute
```

Comment:
Just needs some tests in `test_util.py` in order to fix the coverage. A good test could be to not import a specific dependency (or the module contained inside it) and test some specific function or class that uses that dependency and it should raise an ImportError (which can be checked using `assertRaisesRegex`)

Reply:
I replied to this here: https://github.com/pybamm-team/PyBaMM/pull/3475#discussion_r1387062979